localstack/aws/connect.py

```diff
@@ -0,0 +1,295 @@
+"""
+LocalStack client stack.
+
+This module provides the interface to perform cross-service communication between
+LocalStack providers.
+"""
+import json
+from functools import cache
+from typing import Mapping, Optional, TypedDict, Union
+
+from boto3.session import Session
+from botocore.awsrequest import AWSPreparedRequest
+from botocore.client import BaseClient
+from botocore.config import Config
+from werkzeug.datastructures import Headers
+
+from localstack import config
+from localstack.constants import (
+    INTERNAL_AWS_ACCESS_KEY_ID,
+    INTERNAL_AWS_SECRET_ACCESS_KEY,
+    MAX_POOL_CONNECTIONS,
+)
+from localstack.utils.aws.arns import extract_region_from_arn
+from localstack.utils.aws.aws_stack import get_local_service_url
+from localstack.utils.aws.request_context import get_region_from_request_context
+
+#
+# Data transfer object
+#
+
+LOCALSTACK_DATA_HEADER = "x-localstack-data"
+"""Request header which contains the data transfer object."""
+
+
+class LocalStackData(TypedDict):
+    """
+    LocalStack Data Transfer Object.
+
+    This is sent with every internal request and contains any additional information
+    LocalStack might need for the purpose of policy enforcement. It is serialised
+    into text and sent in the request header.
+
+    Attributes can be added as needed. The keys should roughly correspond to:
+    https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_condition-keys.html
+    """
+
+    source_arn: str
+    """ARN of resource which is triggering the call"""
+
+    source_service: str
+    """Service principal where the call originates, eg. `ec2`"""
+
+    target_arn: str
+    """ARN of the resource being targeted."""
+
+
+def dump_dto(data: LocalStackData) -> str:
+    # TODO@viren: Improve minification using custom JSONEncoder that use shortened keys
+
+    # To produce a compact JSON representation of DTO, remove spaces from separators
+    return json.dumps(data, separators=(",", ":"))
+
+
+def load_dto(data: str) -> LocalStackData:
+    return json.loads(data)
```

Comment:
couldn't we just encode these fields into customer headers of the request? seems that would maybe be more efficient than a JSON codec?

Reply:
It is just simpler to JSON dump into one header than have the logic to parse multiple headers.

However, it is certainly something to consider if we find that our current approach is inefficient for some reason.